artiq/test/coredevice/test_moninj.py

```diff
@@ -66,3 +86,175 @@ def injection_status_cb(channel, override, value):
             (loop_out_channel, TTLOverride.en.value, 1),
             (loop_out_channel, TTLOverride.en.value, 1)
         ])
+
+
+class _UrukulExperiment(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+        
+    @kernel
+    def init_channel(self, channel):
+        self.core.reset()
+        self.core.break_realtime()
+        channel.cpld.init()
+        channel.init()
+
+    @kernel
+    def read_raw(self, channel):
+        self.init_channel(channel)
+        return channel.get_mu()
+
+    @kernel
+    def write_raw(self, channel, ftw: TInt64, pow: TInt32 = 0):
+        self.init_channel(channel)
+        channel.set_mu(ftw, pow)
+
+    @kernel
+    def read(self, channel):
+        self.init_channel(channel)
+        return channel.get()
+
+    @kernel
+    def write(self, channel, freq: TFloat, phase: TFloat = 0.0):
+        self.init_channel(channel)
+        channel.set(freq, phase)
+
+
+class AD991XMonitorTest(ExperimentCase, IsolatedAsyncioTestCase):
+    def get_urukuls(self):
+        urukuls = dict()
+        ddb = self.device_db.get_device_db()
+        for name, desc in ddb.items():
+            if isinstance(desc, dict) and desc["type"] == "local":
+                module, cls = desc["module"], desc["class"]
+                if (module, cls) == ("artiq.coredevice.ad9910", "AD9910"):
+                    urukuls[name] = self.device_mgr.get(name)
+                if (module, cls) == ("artiq.coredevice.ad9912", "AD9912"):
+                    urukuls[name] = self.device_mgr.get(name)
+        return urukuls
+
+    def setUp(self):
+        super().setUp()
+        try:
+            self.core = self.device_mgr.get_desc("core")
+            self.urukuls = self.get_urukuls()
+            self.kernel = self.create(_UrukulExperiment)
+        except KeyError as e:
+            # skip if ddb does not match requirements
+            raise unittest.SkipTest(
+                "test device not available: `{}`".format(*e.args))
+
+    @asynccontextmanager
+    async def open_comm_session(self, notifications=None, core_host=None, auto_monitor=True):
+        if notifications is None:
+            notifications = []
+        if core_host is None:
+            core_host = self.core["arguments"]["host"]
+
+        def monitor_cb(channel, probe, value):
+            notifications.append((channel, probe, value))
+
+        moninj_comm = CommMonInj(monitor_cb, lambda x, y, z: None)
+        try:
+            await moninj_comm.connect(core_host)
+            if auto_monitor:
+                self.setup_monitor(moninj_comm)
+            yield moninj_comm
+        finally:
+            self.setup_monitor(moninj_comm, teardown=True)
+            await moninj_comm._writer.drain()
+            await asyncio.sleep(0.5)
```

Comment:
What the hell?

Reply:
ikr but this is derived from `MonInjTest` above...

Reply:
Where?
If you are talking about this:
```python
moninj_comm.monitor_injection(True, loop_out_channel, TTLOverride.level.en.value)
loop.run_until_complete(asyncio.sleep(0.5))
moninj_comm.inject(loop_out_channel, TTLOverride.level.value, 0)
moninj_comm.inject(loop_out_channel, TTLOverride.level.oe.value, 1)
moninj_comm.inject(loop_out_channel, TTLOverride.level.en.value, 1)
loop.run_until_complete(asyncio.sleep(0.5))
...
```
you are completely misunderstanding what this code does and what the delays are for.

Reply:
Just the fact that there is no delay in the original code between ``drain()`` and ``close()`` should have made alarm bells go off, if you were paying attention.

Reply:
actually I added a sleep there to yield to the asyncio loop for other tasks to continue specifically the receive task, it did not receive anything and failed the test. I changed it to a `asyncio.gather` call on the drain task, should there be better ways?

Reply:
Please pay close attention to the code I copied above and understand it instead of this cargo cult nonsense. Calling ``asyncio.gather`` on a single object is pointless.

Reply:
yeah, it is no better than just `await`ing it. it should be solved now.